ably/realtime_channel.go

```diff
@@ -563,8 +563,10 @@ func (c *RealtimeChannel) notify(msg *proto.ProtocolMessage) {
 		c.setState(ChannelStateFailed, newErrorFromProto(msg.Error))
 		c.queue.Fail(newErrorFromProto(msg.Error))
 	case proto.ActionMessage:
-		for _, msg := range msg.Messages {
-			c.messageEmitter.Emit(subscriptionName(msg.Name), (*subscriptionMessage)(msg))
+		if c.state == ChannelStateAttached {
```

Comment:
This access to `c.state` should happen while holding a lock on `c.mtx`. (This really should be easier to deal with; see #184.)

```suggestion
		c.mtx.Lock()
		state := c.state
		c.mtx.Unlock()
		if state == ChannelStateAttached {
```